### DISCUSSION TITLE
paddleocr  DBnet 输入尺寸问题

### DISCUSSION THREAD

**Discussion by xiaowen0407** [PaddlePaddle/PaddleOCR#13345] [Q&A]
看了下paddleocr2.7的源码 对于DBnet训练网络  他会做EastRandomCropData  尺寸是960 960  ，不够的部分是 填充 ，但是 再做 推理预测时候 是按照最长边 等比例缩放到960  比如 1920 1080的输入图片，那么推理时候尺寸为960，544  而不做填充成 960  960，与训练网络时候的尺寸不一样，这样网络推理的结果还是正确的吗？

**Comment by UserWangZz** [comment]
推理和训练可以输入图像大小不一致。后处理是在分隔图上进行的，训练时resize为960*960是为了提高模型训练的泛化能力，推理时为了得到更好的效果，是按照图像长宽比等比缩放的

**Reply by xiaowen0407** [reply]
谢谢你的答复  可是我还是不懂一点 训练时候 尺寸是960 960 的 换句话说模型学习到的是这个尺寸下的图像信息，尺寸变为960 544 时候模型在960 960 尺寸学习到的知识还能正确在 960 544这个图像上运用吗  输出的概率图的得分还会准确吗？

**Reply by UserWangZz** [reply]
尺度变化是会一定程度影响到模型输出的精度，但是也可以这样理解，不同尺寸的图像resize到960*960本身也是一种增强，因为图像中的信息也相对的发生了变化；而推理阶段我们更关注的是速度，在保证精度的情况下，按照长宽比对图像缩放，图像中的目标没有因为缩放而发生变化，相比于训练也就没有那么难，同时尺度泛化性也是需要考虑的因素。

**Reply by xiaowen0407** [reply]
好的  谢谢哈

**Reply by curryliu606** [reply]
你好请问一下为什么我的识别模型训练出来的结果0.9几，为什么导出推理模型，结果十分差甚至无法识别，我看了一下之前有的讨论，是说训练验证时候做了resize[3,48,320]，但是推理预测时的图片尺寸是动态随机的。那怎么样去解决这个问题呢：1、是在传入预测图片时对图片大小做个resize[3,48,320]吗，这样是可行的吗，还是在检测操作之后对检测出来的部分做resize呢     2、另外的一个看到的方法是，在训练的时候对图片尺寸进行适当修改以匹配预测推理的大小,但是识别模型训练的图片不是通过检测框分割出来的crop_image作为训练和验证数据吗，与推理的原图片尺寸相差较大（我是自己用paddllabel标注的几百张数据作为数据集）

**Reply by curryliu606** [reply]
这是我的配置文件 ch_PP-OCRv3_rec_distillation.yml和ch_PP-OCRv4_rec.yml,ppocrv3和v4都训练了一遍
ch_PP-OCRv3_rec_distillation.yml
Global:
  debug: false
  use_gpu: true
  epoch_num: 500
  log_smooth_window: 20
  print_batch_step: 10
  save_model_dir: ./output/rec_ppocr_v3_distillation
  save_epoch_step: 10
  eval_batch_step: [0, 400]
  cal_metric_during_train: true
  pretrained_model: ./pretrain_models\ch_PP-OCRv3_rec_train\best_accuracy.pdparams
  checkpoints:
  save_inference_dir:
  use_visualdl: false
  infer_img: doc/imgs_words/ch/word_1.jpg
  character_dict_path: ppocr/utils/ppocr_keys_v1.txt
  max_text_length: &max_text_length 25
  infer_mode: false
  use_space_char: true
  distributed: true
  save_res_path: ./output/rec/predicts_ppocrv3_distillation.txt
  d2s_train_image_shape: [3, 48, -1]


Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Piecewise
    decay_epochs : [700]
    values : [0.0005, 0.00005]
    warmup_epoch: 5
  regularizer:
    name: L2
    factor: 3.0e-05


Architecture:
  model_type: &model_type "rec"
  name: DistillationModel
  algorithm: Distillation
  Models:
    Teacher:
      pretrained:
      freeze_params: false
      return_all_feats: true
      model_type: *model_type
      algorithm: SVTR_LCNet
      Transform:
      Backbone:
        name: MobileNetV1Enhance
        scale: 0.5
        last_conv_stride: [1, 2]
        last_pool_type: avg
        last_pool_kernel_size: [2, 2]
      Head:
        name: MultiHead
        head_list:
          - CTCHead:
              Neck:
                name: svtr
                dims: 64
                depth: 2
                hidden_dims: 120
                use_guide: True
              Head:
                fc_decay: 0.00001
          - SARHead:
              enc_dim: 512
              max_text_length: *max_text_length
    Student:
      pretrained:
      freeze_params: false
      return_all_feats: true
      model_type: *model_type
      algorithm: SVTR_LCNet
      Transform:
      Backbone:
        name: MobileNetV1Enhance
        scale: 0.5
        last_conv_stride: [1, 2]
        last_pool_type: avg
        last_pool_kernel_size: [2, 2]
      Head:
        name: MultiHead
        head_list:
          - CTCHead:
              Neck:
                name: svtr
                dims: 64
                depth: 2
                hidden_dims: 120
                use_guide: True
              Head:
                fc_decay: 0.00001
          - SARHead:
              enc_dim: 512
              max_text_length: *max_text_length
Loss:
  name: CombinedLoss
  loss_config_list:
  - DistillationDMLLoss:
      weight: 1.0
      act: "softmax"
      use_log: true
      model_name_pairs:
      - ["Student", "Teacher"]
      key: head_out
      multi_head: True
      dis_head: ctc
      name: dml_ctc
  - DistillationDMLLoss:
      weight: 0.5
      act: "softmax"
      use_log: true
      model_name_pairs:
      - ["Student", "Teacher"]
      key: head_out
      multi_head: True
      dis_head: sar
      name: dml_sar
  - DistillationDistanceLoss:
      weight: 1.0
      mode: "l2"
      model_name_pairs:
      - ["Student", "Teacher"]
      key: backbone_out
  - DistillationCTCLoss:
      weight: 1.0
      model_name_list: ["Student", "Teacher"]
      key: head_out
      multi_head: True
  - DistillationSARLoss:
      weight: 1.0
      model_name_list: ["Student", "Teacher"]
      key: head_out
      multi_head: True

PostProcess:
  name: DistillationCTCLabelDecode
  model_name: ["Student", "Teacher"]
  key: head_out
  multi_head: True

Metric:
  name: DistillationMetric
  base_metric_name: RecMetric
  main_indicator: acc
  key: "Student"
  ignore_space: False

Train:
  dataset:
    name: SimpleDataSet
    data_dir: ./train_data/rec/train
    ext_op_transform_idx: 1
    label_file_list:
    - ./train_data/rec/train.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - RecConAug:
        prob: 0.5
        ext_data_num: 2
        image_shape: [48, 320, 3]
        max_text_length: *max_text_length
    - RecAug:
    - MultiLabelEncode:
    - RecResizeImg:
        image_shape: [3, 48, 320]
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_sar
        - length
        - valid_ratio
  loader:
    shuffle: true
    batch_size_per_card: 32
    drop_last: true
    num_workers: 0
Eval:
  dataset:
    name: SimpleDataSet
    data_dir: ./train_data/rec/val
    label_file_list:
    - ./train_data/rec/val.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - MultiLabelEncode:
    - RecResizeImg:
        image_shape: [3, 48, 320]
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_sar
        - length
        - valid_ratio
  loader:
    shuffle: false
    drop_last: false
    batch_size_per_card: 32
    num_workers: 0


ch_PP-OCRv4_rec.yml
Global:
  debug: false
  use_gpu: true
  epoch_num: 500
  log_smooth_window: 20
  print_batch_step: 10
  save_model_dir: ./output/rec_ppocr_v4
  save_epoch_step: 10
  eval_batch_step: [0, 200]
  cal_metric_during_train: true
  pretrained_model: ./pretrain_models\ch_PP-OCRv4_rec_train\best_accuracy.pdparams
  checkpoints:
  save_inference_dir:
  use_visualdl: false
  infer_img: doc/imgs_words/ch/word_1.jpg
  character_dict_path: ppocr/utils/ppocr_keys_v1.txt
  max_text_length: &max_text_length 25
  infer_mode: false
  use_space_char: true
  distributed: true
  save_res_path: ./output/rec/predicts_ppocrv4.txt


Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Cosine
    learning_rate: 0.001
    warmup_epoch: 5
  regularizer:
    name: L2
    factor: 3.0e-05


Architecture:
  model_type: rec
  algorithm: SVTR_LCNet
  Transform:
  Backbone:
    name: PPLCNetV3
    scale: 0.95
  Head:
    name: MultiHead
    head_list:
      - CTCHead:
          Neck:
            name: svtr
            dims: 120
            depth: 2
            hidden_dims: 120
            kernel_size: [1, 3]
            use_guide: True
          Head:
            fc_decay: 0.00001
      - NRTRHead:
          nrtr_dim: 384
          max_text_length: *max_text_length

Loss:
  name: MultiLoss
  loss_config_list:
    - CTCLoss:
    - NRTRLoss:

PostProcess:  
  name: CTCLabelDecode

Metric:
  name: RecMetric
  main_indicator: acc

Train:
  dataset:
    name: MultiScaleDataSet
    ds_width: false
    data_dir: ./train_data/rec/train
    ext_op_transform_idx: 1
    label_file_list:
    - ./train_data/rec/train.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - RecConAug:
        prob: 0.5
        ext_data_num: 2
        image_shape: [48, 320, 3]
        max_text_length: *max_text_length
    - RecAug:
    - MultiLabelEncode:
        gtc_encode: NRTRLabelEncode
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_gtc
        - length
        - valid_ratio
  sampler:
    name: MultiScaleSampler
    scales: [[320, 32], [320, 48], [320, 64]]
    first_bs: &bs 32
    fix_bs: false
    divided_factor: [8, 16] # w, h
    is_training: True
  loader:
    shuffle: true
    batch_size_per_card: *bs
    drop_last: true
    num_workers: 8
Eval:
  dataset:
    name: SimpleDataSet
    data_dir: ./train_data/rec/val
    label_file_list:
    - ./train_data/rec/val.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - MultiLabelEncode:
        gtc_encode: NRTRLabelEncode
    - RecResizeImg:
        image_shape: [3, 48, 320]
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_gtc
        - length
        - valid_ratio
  loader:
    shuffle: false
    drop_last: false
    batch_size_per_card: 32
    num_workers: 0

**Reply by curryliu606** [reply]
有没有什么好的解决办法大佬求解

**Reply by xiaowen0407** [reply]
resize[3,48,320]   
传入识别模型前 图片一定要是这种形状的  我是C++ 部署的 我看源码 在调用识别模型前 是进行 resize到这个尺寸的 不够的地方 padding了 的

**Reply by UserWangZz** [reply]
训练的时候是动态分辨率训练，推理验证的时候是固定3*48*320推理的

**Reply by xiaowen0407** [reply]
对  推理时候 要固定到  348320推理

**Reply by curryliu606** [reply]
> resize[3,48,320] 传入识别模型前 图片一定要是这种形状的 我是C++ 部署的 我看源码 在调用识别模型前 是进行 resize到这个尺寸的 不够的地方 padding了 的

所以还是在识别前进行个resize吗，所以我看那个utility设置的rec_img_shape是3,48,320，必须得是这个尺寸，但是我传入的识别图片不是这个尺寸，所以才导致效果差吗

 # params for text recognizer
    parser.add_argument("--rec_algorithm", type=str, default="SVTR_LCNet")
    parser.add_argument("--rec_model_dir", type=str,default="./output/inference/ch_PP-OCR_V3_rec_infer/Student")
    parser.add_argument("--rec_image_inverse", type=str2bool, default=True)
    parser.add_argument("--rec_image_shape", type=str, default="3, 48, 320")
    parser.add_argument("--rec_batch_num", type=int, default=6)
    parser.add_argument("--max_text_length", type=int, default=25)
    parser.add_argument(
        "--rec_char_dict_path", type=str, default="./ppocr/utils/ppocr_keys_v1.txt"
    )
    parser.add_argument("--use_space_char", type=str2bool, default=True)
    parser.add_argument("--vis_font_path", type=str, default="./doc/fonts/simfang.ttf")
    parser.add_argument("--drop_score", type=float, default=0.5)

**Reply by curryliu606** [reply]
所以我是需要在predict_det检测之前做缩放，还是predict_rec之前做呢
因为我直接是运行的tools/infer/predict_system.py做串联推理的

**Reply by curryliu606** [reply]
因为，我直接用官方的ocrv4的识别推理模型就能正常推理识别出来，也不需要做别的操作，所以搞不懂是不是图片的问题，那官方的为啥不用做这个也有很好的识别效果的呢

**Reply by UserWangZz** [reply]
是先det然后crop出文本区域，对文本区域会resize然后识别

**Reply by curryliu606** [reply]
> 是先det然后crop出文本区域，对文本区域会resize然后识别
  img_crop_list = []

        dt_boxes = sorted_boxes(dt_boxes)

        for bno in range(len(dt_boxes)):
            tmp_box = copy.deepcopy(dt_boxes[bno])
            tmp_box[0][0]-=20
            tmp_box[0][1]-=20
            tmp_box[1][0]+=20
            tmp_box[1][1]-=20
            tmp_box[2][0]+=20
            tmp_box[2][1]+=20
            tmp_box[3][0]-=20
            tmp_box[3][1]+=20
            if self.args.det_box_type == "quad":
                img_crop = get_rotate_crop_image(ori_im, tmp_box)
            else:
                img_crop = get_minarea_rect_crop(ori_im, tmp_box)

            img_crop = cv2.resize(img_crop, (320, 48), interpolation= cv2.INTER_LINEAR)  
            img_crop_list.append(img_crop)


这样缩放吗，但是更差了啊，本来直接下载的官网模型可以识别的，压缩成这样都识别不出来了，我看了一下做这个resize后文本内容压缩的都变形了，很难识别
